### ISSUE TITLE
Automated Changelog

### ISSUE THREAD

**Issue by dOrgJelli** [Web3-API/prototype#282]
Automate some pieces of the changelog: contributors, PRs / features, etc

**Comment by pwvpwvpwv** [comment]
Creating the changelog is a tedious manual process that is prone to human error. However, creating a changelog whenever a new version is going to be released is highly beneficial to any open source project, but especially so in a project with as many components as Polywrap: changelogs inform contributors of the current state of the project and help them keep track of what's going on in various parts of the codebase; they also allow potential adopters and contributors to gain some insight into how quickly the project is changing and what direction its going in. Therefore, developing a system to automate the changelog process that ties into the development cycle as a whole and enforces helpful and descriptive commit messages is advantageous on all spectrums. 

## Method
Automating the changelog utilizing *Git History*, *SemVer* and *Conventional Commit* formatting is the most common combination of tools people use to solve this problem. These tools are used in conjunction with a library that allows for Git hooks in order to create a quick, clean, and standardized commit messaging protocol that can drive the content and automation of the changelog. This is a process we've previously seen employed to great effect, and from experience believe it should be instituted in projects earlier rather than later for best results. The policy changes help to standardize efforts across a distributed and/or asynchronous development team, and the software components help to make the process easier while keeping some of the human error out of the equation.

### Semantic Versioning
Semantic Versioning (SemVer) is a standard for code versioning. It specifies that a version number should always be monotonically increasing and contain these three parts:

 - **MAJOR**: incremented when you add breaking changes, *e.g.* an incompatible API change
 - **MINOR**: incremented when you add backward compatible functionality
 - **PATCH**: incremented when you add backward compatible bug fixes
 
### Conventional Commit
The Conventional Commits specification describes a standardized, lightweight convention on top of commit messages. This convention dovetails with SemVer, asking software developers to describe in commit messages any features, fixes, and breaking changes that they make. Adopting the Conventional Commits specification will discourage/prevent commit messages that do not describe where changes were made, what was changed, and what was the motivation for making the changes (all useful things that help reduce friction within a team and make it easy for others to quickly digest new or proposed changes).

```
<type>[optional scope]: <description>

[optional body]

[optional footer]
```

An example of such a message:

```
fix: ABC-123: Caught Promise exception

We did not catch the promise exception thrown by the API call
and therefore we could not show the error message to the user
```

### Implementation Details
While there are many tools out there to help automate and enforce these new policies, we believe that Polywrap will find success quickest by: using [`@commitlint/config-conventional`](https://github.com/conventional-changelog/commitlint/tree/master/%40commitlint/config-conventional) to enforce Conventional Commits via Git hook; and using [`standard-version`](https://github.com/conventional-changelog/standard-version), a utility for versioning using SemVer and changelog generation powered by Conventional Commits.

#### Example flow
Suppose you are a first-time contributor who's just written a new bit of functionality for the Polywrap JS client. Now you need to install and engage with some tooling to help automate your commit(s) changelog and ensure it provides the information necessary to help others in the project better understand your changes.

First, install [`husky`](https://github.com/typicode/husky), which sets up Git hooks:
```
npx husky-init && npm install
```

Then, install `commitlint` with a config, which will be used to lint your commit message:
```
npm install @commitlint/{cli,config-conventional}
```

After installing `standard-version`, we can run the appropriate script that Polywrap has defined in the `package.json`:
```json
// ...

"scripts": {
    "release":       "standard-version",
    "release:minor": "standard-version --release-as minor",
    "release:patch": "standard-version --release-as patch",
    "release:major": "standard-version --release-as major"
},

// ...
```

Polywrap can configure the changelog generation to suit the project's unique needs via a `.versionrc.json` file or placing a `standard-version` stanza in `package.json`:
```json
{
    "types": [
        {"type": "feat", "section": "Features"},
        {"type": "fix", "section": "Bug Fixes"},
        {"type": "chore", "hidden": true},
        {"type": "docs", "hidden": true},
        {"type": "style", "hidden": true},
        {"type": "refactor", "hidden": true},
        {"type": "perf", "hidden": true},
        {"type": "test", "hidden": true}
    ],
    "commitUrlFormat": "https://github.com/mokkapps/changelog-generator-demo/commits/{{hash}}",
    "compareUrlFormat": "https://github.com/mokkapps/changelog-generator-demo/compare/{{previousTag}}...{{currentTag}}"
}
```

## Additional Thoughts
These policies and changes should be codified and made explicit in a `CONTRIBUTING.md` (or similarly named file) at the top of the Polywrap repo to ensure all contributors are aware of whats expected of them.

The proposed method would require the addition of four dev dependencies to the project. However, these are powerful tools that are regularly maintained, with strong public backing, and that will allow for a smoother, less tedious commit and changelog process. Having a streamlined, standardized changelog process will encourage contributors to write better commit messages, as well as be able to rely on the commit messages of others when they are attempting to understand part of the codebase.

**Comment by evanjacobs** [comment]
I agree with most of the proposal here but I have one question. There is often additional context about a bug or feature in the discussion section of a Github issue. What is the best way to link from the changelog to the related issue that is fixed in the release?

**Comment by Niraj-Kamdar** [comment]
@pwvpwvpwv  I agree with using conventional commit and semantic versioning but one issue I have is: while working on a PR, we may be touching various parts of the code bases and having atomic commits that describe what we changed may be a lots of bookkeeping from dev side. I think a good balance would be to having atomic PR with the conventional commit and changelog so we can summarise changes on PR bases once we complete the feature implementation in PR description.

**Comment by pwvpwvpwv** [comment]
@Niraj-Kamdar I think the onus for a developer to follow the conventional commit messages is relatively minor. A commit message simply requires a type and a title in the following format: `{type}: {title}` for the commit. 

Also, developers can use `--no-verify` flag to ignore the commit hook lint of the commit message for intermediate commits that wouldn't be part of the changelog. These are not linted, as it's only the current commit you are creating a message for that is linted. So they could theoretically use that for any commits that aren't intended to be included in the changelog, with the final commit message being the one that includes the required information. This seems to allow maximum flexibility, as any commit messages that are appropriate for the changelog should be in this format and linted and I can use the `--no-verify` flag to indicate when a commit should not be included.

Let me know if this is still a concern - if we want to base it purely on the PR message submitted we might need to run this in a different way and implement some changes to the GitHub workflow that is set up.

**Comment by Niraj-Kamdar** [comment]
It makes sense! But would love to hear opinions of other devs.

**Comment by krisbitney** [comment]
I agree with Niraj that it may be better to base the changelog on PRs. 

It's simple to use the `{type}: {title}` format for messages, and I think it is a good idea for us adopt to this practice.

But I'm not sure that using commit messages to automate the changelog is as simple as using a new message format. A single new feature can be implemented in pieces across several commits, sometimes dozens. For a new feature would we use the `feat` label in the first commit of the branch? Would using commit messages for the changelog incentivize different commit patterns?

**Comment by nerfZael** [comment]
I agree with Niraj's point about unnecessary bookkeeping. It does feel like, to me, that using commits in this way is stretching the git commit functionality and appears "hacky". Especially with what Kris said with features and fixes being split across several commits. Using a final commit to explain the changes doesn't seem like the best use of commits. 
While working on a branch there can be multiple times a fix or a mini-feature is added/removed and documenting that is redundant. 
I'm of the opinion that documenting changes in the PR and hooking that up to the changelog is the way to go. This would also have the added benefit of reviewers being able to add additional changelog messages if the author of the PR missed any. (Something that can't be done with commits)

**Comment by pwvpwvpwv** [comment]
Pulling this from discord so it's in both places for potential viewers:

> So, one of the main points of discussion here seems to be at what level the changelog automation should be taking place. Here's a quick snippet from the Conventional Commits website:
> 
> > **Do all my contributors need to use the Conventional Commits specification?**
> > No! If you use a squash based workflow on Git lead maintainers can clean up the commit messages as they’re merged—adding no workload to casual committers. A common workflow for this is to have your git system automatically squash commits from a pull request and present a form for the lead maintainer to enter the proper git commit message for the merge.
> 
> What does the merging strategy look like on the Polywrap project? If PRs are squash-merged then it's simply the responsibility of the person handling the merge to update the commit message to be pulled into the changelog which is generated during the GitHub workflow. 
> 

Any thoughts on this approach - specifically the squashing on merge strategy? From my quick review it looks like there isn't a squash being used currently.

**Comment by nerfZael** [comment]
I think that would be good if we had a squash merge practice in place, but we don't.
We could start practicing squash merge, but I personally have some reservations.
E.g. with squash merge you lose the implementation reasons that you normally find in commits and it makes it harder for devs looking at git history/git blame to find out why something was done the way it was, which, in my opinion, is the main benefit of commit messages in the first place.
It just seems to me like commits are a bad vehicle for changelog, and, while they can be used for that, add more cons than pros.
Take this opinion with a grain of salt, I'm sure there are many projects successfully using them for this purpose.

An alternative approach could be to have the burden be on the PR Author to add a changelog comment to the PR. The reviewers can see the changelog and ask for revisions if necessary.
When a branch is being published, one would need to then go through the PRs that made it into the release and extract the changelog and place it where appropriate (web, forum, discord, etc)
This could also be automated to an extent

**Comment by Niraj-Kamdar** [comment]
I saw many projects using squash and merge or rebase and main reason for this is to keep the linear history but there is a good article which explains why rewriting the git history isn't a good idea: https://medium.com/@fredrikmorken/why-you-should-stop-using-git-rebase-5552bee4fed1

This mainly focus on rebase but I think same could be said for squash and merge since it also replaces original git commits with single atomic commit.

**Comment by dOrgJelli** [comment]
I think that in order to move forward with this, it'd be best to just use the PR to encode updates to the changelog. In the future we can discuss the potential of using squash merge or standardized commit messaging if we feel it's necessary.
